### ISSUE TITLE
Add a blog post about using the verifier to help write an indexer. [phab:D400]

### ISSUE THREAD

**Issue by kythe** [google/kythe#630]
_Created by **zarko** at 2015-07-14 22:20:55_: 

Blocking on #612 (since that adds a command line flag to
change the verifier comment prefix that the post mentions.)

**Comment by kythe** [comment]
_**fromberger** commented inline (context lost during migration) at 2015-07-21 15:53:01_:

I'd move this line below the (current) next paragraph:

> In the previous article …
> 
> Now we'll use the Kythe verifier to…

**Comment by kythe** [comment]
_**fromberger** commented inline (context lost during migration) at 2015-07-21 15:53:01_:

I think more context would be helpful here, and that "goals" should be more explicitly defined. Many of these ideas aren't in the common traffic patterns for most developers.

Perhaps:

> Given a description of the anchors and their desired relationships, performing these checks doesn't require any information specific to the language being indexed. With this in mind, we built the Kythe verifier. The verifier accepts a stream of Kythe entries and source files, the latter of which have been annotated with *goals*. Each goal describes entries that the verifier must (or must not) …

**Comment by kythe** [comment]
_**fromberger** commented inline (context lost during migration) at 2015-07-21 15:53:01_:

Suggestion: 

> The `//-` prefix tells the verifier…

**Comment by kythe** [comment]
_**fromberger** commented inline (context lost during migration) at 2015-07-21 15:53:01_:

Maybe summarize what these entries are?

> Here, we are providing JSON-encoded entries representing the kind (`file`) and contents (`Hello, world!`) of our example: …

**Comment by kythe** [comment]
_**fromberger** commented inline (context lost during migration) at 2015-07-21 15:53:01_:

> …an unsatisfiable goal—for example, a node with kind `elif`:

**Comment by kythe** [comment]
_**fromberger** commented inline (context lost during migration) at 2015-07-21 15:53:01_:

I'd flip the order here: If you're not already…you'll need to…

**Comment by kythe** [comment]
_**fromberger** commented inline (context lost during migration) at 2015-07-21 15:53:01_:

This seems sudden. It's valuable to show how to get the graph, but a bit more exposition might help.

> …the verifier won't have any goals to check, and will (trivially) succeed.
>
> Even without any goals, however, we can get some useful information from the verifier: By passing the `--annotated_graphviz` flag, we ask the verifier to emit a Graphviz graph representing all the nodes and relationships found in the input:
>
> [script]
>
> This will render in xdot as something like:

**Comment by kythe** [comment]
_**fromberger** commented inline (context lost during migration) at 2015-07-21 15:53:01_:

> … the special `//-` prefix:

**Comment by kythe** [comment]
_**fromberger** commented inline (context lost during migration) at 2015-07-21 15:53:01_:

This could bear with a couple lines of explanation for what it means (I don't think it needs to be complete or formal at this point, just an intuition for what the goal notation is—variables, labels).

**Comment by kythe** [comment]
_**fromberger** commented inline (context lost during migration) at 2015-07-21 15:53:01_:

Suggestion:

> … with its use in the `return` statement.

"Mention" is correct but stilted, and this paragraph is already pretty packed.

**Comment by kythe** [comment]
_**fromberger** commented inline (context lost during migration) at 2015-07-21 15:53:01_:

> … let's instead limit the search to only anchor nodes that identify the specific source text range we really want:

**Comment by kythe** [comment]
_**fromberger** commented inline (context lost during migration) at 2015-07-21 15:53:01_:

Maybe summarize the `` notation somewhere nearby.

**Comment by kythe** [comment]
_**fromberger** wrote at 2015-07-21 15:53:01_:

Overall I think this is quite good. I made a bunch of comments, most of which are intended to be suggestions rather than demands. :)
There are a few places where I think a bit more context would be helpful, noted in the comments.

**Comment by kythe** [comment]
_**schroederc** commented inline (context lost during migration) at 2015-07-21 16:08:28_:

This is the tag line for the post.  It'll show up on the main blog page: http://www.kythe.io/blog

**Comment by kythe** [comment]
_**schroederc** commented inline (context lost during migration) at 2015-07-21 16:08:28_:

This makes it seem that we only care about cross-references.  It's more of the case that our verifier uses anchors as its start point in the graph.  It can then test the rest of the relationships transitively.

Can we make it clear that we don't **just** care about the anchors?

**Comment by kythe** [comment]
_**schroederc** commented inline (context lost during migration) at 2015-07-21 16:08:28_:

Maybe a definition of a stream?  Or a footnote?  Or link the word "stream" to http://godoc.org/kythe.io/kythe/go/platform/delimited?

**Comment by kythe** [comment]
_**schroederc** commented inline (context lost during migration) at 2015-07-21 16:08:28_:

I'd like to replace all uses of Bazel with the releases: https://github.com/google/kythe/releases.

We should add the verifier to the package.

**Comment by kythe** [comment]
_**schroederc** required changes at 2015-07-21 16:08:28_

**Comment by kythe** [comment]
_**schroederc** commented inline (context lost during migration) at 2015-07-21 16:13:41_:

Just `entrystream --read_json`

**Comment by kythe** [comment]
_**schroederc** commented inline (context lost during migration) at 2015-07-21 16:13:41_:

And Java?

**Comment by kythe** [comment]
_**schroederc** commented inline (context lost during migration) at 2015-07-21 16:13:41_:

Save this in another file?

**Comment by kythe** [comment]
_**schroederc** commented inline (context lost during migration) at 2015-07-21 16:13:41_:

Save in another file?

**Comment by kythe** [comment]
_**schroederc** commented inline (context lost during migration) at 2015-07-21 16:13:41_:

While we're here, please add a footnote or paragraph noting that these binaries are prebuilt at https://github.com/google/kythe/releases

**Comment by kythe** [comment]
_**zarko** commented inline (context lost during migration) at 2015-07-22 00:31:34_:

Do we also plan on cutting releases for OS X?

**Comment by kythe** [comment]
_**zarko** commented inline (context lost during migration) at 2015-07-22 00:31:34_:

Is the new version what you had in mind?

**Comment by kythe** [comment]
_**zarko** commented inline (context lost during migration) at 2015-07-22 00:31:35_:

I'm leaving this be because of the functionality Cody described.

**Comment by kythe** [comment]
_**zarko** wrote at 2015-07-22 00:33:00_:

  - Address comments.

**Comment by kythe** [comment]
_**zarko** wrote at 2015-08-05 17:48:25_:

ping

**Comment by kythe** [comment]
_**schroederc** commented inline (context lost during migration) at 2015-08-05 17:58:21_:

I'm not sure where this has been "done".

**Comment by kythe** [comment]
_**schroederc** commented inline (context lost during migration) at 2015-08-05 17:58:21_:

We can definitely do OSX, if we wanted.  I've just not heard of any demand.

The release package now has the verifier, so I don't see any reason to use Bazel in the blog posts anymore.  I think it just adds unnecessary complexity for those following along.

**Comment by kythe** [comment]
_**schroederc** commented inline (context lost during migration) at 2015-08-05 17:58:21_:

"$"

**Comment by kythe** [comment]
_**schroederc** commented inline (context lost during migration) at 2015-08-05 17:58:21_:

s/or/and/

**Comment by kythe** [comment]
_**schroederc** commented inline (context lost during migration) at 2015-08-05 17:58:21_:

Can we put these in `_posts/imgs/` or something similiar for more organization?

**Comment by kythe** [comment]
_**zarko** wrote at 2015-08-05 19:52:23_:

  - Address comments.
  - Address comments.

**Comment by kythe** [comment]
_**schroederc** commented inline (context lost during migration) at 2015-08-05 19:59:25_:

Eh?

**Comment by kythe** [comment]
_**zarko** commented inline (context lost during migration) at 2015-08-05 20:08:01_:

Below, in a comment in the script that explains why we're using the `entrystream` tool. Understanding the particulars of the encoding isn't important  at this point in the document.

**Comment by kythe** [comment]
_**zarko** commented inline (context lost during migration) at 2015-08-05 20:08:01_:

There's no need to talk about protocol buffers in this part of the explanation. Interested readers can look at the script, where the extra step to go from JSON to (delimited) proto is described as:

```
# Read JSON entries from standard in and pass them to the verifier.
# The entrystream tool turns the JSON into length-delimited protocol buffers,
# described at http://godoc.org/kythe.io/kythe/go/platform/delimited
/opt/kythe/tools/entrystream --read_json | /opt/kythe/tools/verifier "$
```

**Comment by kythe** [comment]
_**schroederc** accepted this revision at 2015-08-05 20:10:08_

**Comment by kythe** [comment]
_**zarko** wrote at 2015-08-05 20:26:20_:

  - Update date.

**Comment by kythe** [comment]
_**zarko** closed via commit 0246eece6ed39ca4c247337db12e9d25cc8c248d at 2015-08-05 20:26:47_
